text/2540-formal-function-parameter-attributes.md

```diff
@@ -0,0 +1,242 @@
+
+## Errors & Warnings
+
+There are two errornous situations when it comes to 
+
+### Warning: Unused attribute
+
+When using an non-defined attribute that is not used by either the language or a custom defined procedural macro.
+
+```
+warning: unused attribute
+ --> src/main.rs:2
+  |
+2 | #[foo] bar: u32
+  | ^^^^^^^^^
+  |
+  = note: #[warn(unused_attributes)] on by default
+```
+
+### Error: Malformed attribute
+
+When using a known attribute that is not defined for formal parameters such as when attributing `fn main` with `#[allow]`.
+
+Example shows the usage of the known attribute `#[inline]` used on a formal parameter without being defined for it.
+
+```
+error[E0452]: malformed lint attribute
+ --> src/main.rs:2
+  |
+2 | #[inline] bar: u32
+  | ^^^^^^^^
+```
+
+The same applies for attributes with an incorrect format such as `#[inline(key = value)]` that is handled as its done in other contexts.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Description
+
+In accordance to the RFC for [attributes for generic params](https://github.com/frol/rust-rfcs/blob/master/text/1327-dropck-param-eyepatch.md)
+this feature is guarded by the `formal_function_param_attrs` feature guard.
+
+The grammar of the following language items has to be adjusted to allow that
+constructions like the following will become legal.
+
+- Function definitions
+- Method definitions
+- Trait function declarations and defnitions (with or without default impl)
+- Lambda & closure definitions
+
+### Example: A single attributed parameter for function decl or definition
+
+```rust
+fn foo(#[bar] baz: bool);
+fn bar(#[bar] qux: bool) { println!("hi"); }
+```
+
+### Example: For methods or trait function definitions
+
+```rust
+fn into_foo(#[bar] self);
+fn foo(#[bar] &self);
+fn foo_mut(#[bar] &mut self);
+```
+
+### Example: Multiple attributed parameters
+
+```rust
+// Twice the same attribute
+fn fst_foo(#[bar] baz: bool, #[bar] qiz: u32);
+
+// Different attributes
+fn snd_foo(#[bar] baz: bool, #[qux] qiz: u32);
+```
+
+### Example: Any structured attribute
+
+```rust
+fn foo(#[bar(Hello)] baz: bool);
+fn bar(#[qux(qiz = World)] baz: bool);
+```
+
+### Example: Lambdas & closures
+
+```rust
+let mut v = [5, 4, 1, 3, 2];
+v.sort_by(|#[bar] a, #[baz] b| a.cmp(b));
+```
+
+## Errors
+
+Users can encounter two different errorneous situations.
+
+### Unknown attribute used
+
+When a user is using an attribute that is not known at the point of its invokation
+a warning is generated similar to other usages of unknown attributes in the language.
+
+This may for example happen in the context of a procedural macros.
+
+### Malformed attribute
+
+When a user is using a known or language defined attribute at a non supported location
+an error is generated like in other usages of malformed attributes in the language.
+An example can be seen in the previous section.
+
```

Comment:
It is unclear to me whether `#[doc ...]` is permitted by this RFC... should we allow?:
```rust
fn foo(
    /// some docs...
    arg: usize
) {
    ...
}
```

Reply:
In my opinion this would be awesome!
At least I had the idea that it would be so much better than having to reference parameters in the function doc if there is something special to note about them.

However, I am aware that this could lead to yet another round of unresolved questions for the doc generation.

Reply:
I have integrated a small section about doc comments in the motivation.